### ISSUE TITLE
Unable to pip install on OSX

### ISSUE THREAD

**Issue by Karsten1987** [robotology/idyntree#876]
We're trying out the library in the context of ros2_control but having trouble to install the library through pip. Locally on my OSX with pyenv 3.8 I get quite a verbose output when trying to install the package via pip - apparently getting confused with three different python interpreters. 

Is the pip package supported on OSX?

```
 ➭ python --version
Python 3.8.0
```

```
 ➭ pip install idyntree
Collecting idyntree
  Using cached idyntree-3.2.1.tar.gz (1.2 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: numpy in ./.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/site-packages (from idyntree) (1.20.1)
Building wheels for collected packages: idyntree
  Building wheel for idyntree (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/bin/python3.8 /Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/tmpetki7_03
       cwd: /private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-install-xinkxb_o/idyntree_6e7a01ce97a940e69151d9a2f521d9c0
  Complete output (124 lines):
  running bdist_wheel
  running build
  running build_ext
  -- The C compiler identification is Clang 10.0.1
  -- The CXX compiler identification is Clang 10.0.1
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/local/opt/ccache/libexec/clang - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/local/opt/ccache/libexec/clang++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Found LibXml2: /usr/lib/libxml2.dylib (found version "2.9.4")
  -- Found IPOPT: /usr/local/lib/libipopt.dylib
  CMake Warning (dev) at /usr/local/lib/cmake/assimp-5.0/assimpTargets.cmake:54 (if):
    if given arguments:

      "ON"

    An argument named "ON" appears in a conditional statement.  Policy CMP0012
    is not set: if() recognizes numbers and boolean constants.  Run "cmake
    --help-policy CMP0012" for policy details.  Use the cmake_policy command to
    set the policy and suppress this warning.
  Call Stack (most recent call first):
    /usr/local/lib/cmake/assimp-5.0/assimp-config.cmake:1 (include)
    cmake/iDynTreeDependencies.cmake:55 (find_package)
    cmake/iDynTreeDependencies.cmake:97 (idyntree_handle_dependency)
    CMakeLists.txt:39 (include)
  This warning is for project developers.  Use -Wno-dev to suppress it.

  -- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
  -- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
  -- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
  -- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
  -- Performing Test COMPILER_HAS_DEPRECATED_ATTR
  -- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success
  -- Found SWIG: /usr/local/bin/swig (found version "4.0.2")
  -- Found Python3: /usr/local/Frameworks/Python.framework/Versions/3.9/bin/python3.9 (found version "3.9.1") found components: Interpreter Development.Module NumPy
  -- Performing Test CXX_HAS_WNO_DEPRECATED_DECLARATIONS
  -- Performing Test CXX_HAS_WNO_DEPRECATED_DECLARATIONS - Success
  -- Performing Test CXX_HAS_WNO_DEPRECATED
  -- Performing Test CXX_HAS_WNO_DEPRECATED - Success
  -- Found PythonInterp: /Users/karsten/.pyenv/shims/python3.7
  CMake Error at /usr/local/share/cmake/pybind11/FindPythonLibsNew.cmake:95 (message):
    Python config failure:

    pyenv: python3.7: command not found



    The `python3.7' command exists in these Python versions:

      3.7.0
      3.7.4



    Note: See 'pyenv help global' for tips on allowing both

          python2 and python3 to be found.

  Call Stack (most recent call first):
    /usr/local/share/cmake/pybind11/pybind11Tools.cmake:16 (find_package)
    /usr/local/share/cmake/pybind11/pybind11Config.cmake:100 (include)
    bindings/CMakeLists.txt:100 (find_package)


  -- Configuring incomplete, errors occurred!
  See also "/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-install-xinkxb_o/idyntree_6e7a01ce97a940e69151d9a2f521d9c0/build/temp.macosx-10.14-x86_64-3.8_BuildAndInstall/CMakeFiles/CMakeOutput.log".

  ==> Configuring:
  $ cmake -S /private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-install-xinkxb_o/idyntree_6e7a01ce97a940e69151d9a2f521d9c0 -B /private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-install-xinkxb_o/idyntree_6e7a01ce97a940e69151d9a2f521d9c0/build/temp.macosx-10.14-x86_64-3.8_BuildAndInstall -GNinja -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX:PATH=/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-install-xinkxb_o/idyntree_6e7a01ce97a940e69151d9a2f521d9c0/build/lib.macosx-10.14-x86_64-3.8/idyntree -DBUILD_SHARED_LIBS:BOOL=OFF -DIDYNTREE_USES_PYTHON:BOOL=ON -DIDYNTREE_COMPILE_TESTS:BOOL=OFF -DIDYNTREE_PACKAGE_FOR_PYPI:BOOL=ON -DIDYNTREE_USES_IPOPT:BOOL=ON -DIDYNTREE_USES_ASSIMP:BOOL=ON -DIDYNTREE_USES_IRRLICHT:BOOL=OFF -DIDYNTREE_USES_QT5:BOOL=OFF -DIDYNTREE_USES_OSQPEIGEN:BOOL=OFF -DIDYNTREE_USES_ALGLIB:BOOL=OFF -DIDYNTREE_USES_WORHP:BOOL=OFF -DIDYNTREE_USES_YARP:BOOL=OFF -DIDYNTREE_USES_ICUB_MAIN:BOOL=OFF

  ==> Building:
  $ cmake --build /private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-install-xinkxb_o/idyntree_6e7a01ce97a940e69151d9a2f521d9c0/build/temp.macosx-10.14-x86_64-3.8_BuildAndInstall --config Release

  ==> Installing:
  $ cmake --install /private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-install-xinkxb_o/idyntree_6e7a01ce97a940e69151d9a2f521d9c0/build/temp.macosx-10.14-x86_64-3.8_BuildAndInstall

  Traceback (most recent call last):
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 280, in <module>
      main()
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 263, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 204, in build_wheel
      return _build_backend().build_wheel(wheel_directory, config_settings,
    File "/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-build-env-nt4fesab/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 221, in build_wheel
      return self._build_with_temp_dir(['bdist_wheel'], '.whl',
    File "/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-build-env-nt4fesab/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 207, in _build_with_temp_dir
      self.run_setup()
    File "/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-build-env-nt4fesab/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 150, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 15, in <module>
      setup(
    File "/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-build-env-nt4fesab/overlay/lib/python3.8/site-packages/setuptools/__init__.py", line 153, in setup
      return distutils.core.setup(**attrs)
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/distutils/core.py", line 148, in setup
      dist.run_commands()
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/distutils/dist.py", line 966, in run_commands
      self.run_command(cmd)
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/distutils/dist.py", line 985, in run_command
      cmd_obj.run()
    File "/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-build-env-nt4fesab/overlay/lib/python3.8/site-packages/wheel/bdist_wheel.py", line 299, in run
      self.run_command('build')
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/distutils/dist.py", line 985, in run_command
      cmd_obj.run()
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/distutils/command/build.py", line 135, in run
      self.run_command(cmd_name)
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/distutils/dist.py", line 985, in run_command
      cmd_obj.run()
    File "/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-build-env-nt4fesab/overlay/lib/python3.8/site-packages/cmake_build_extension/build_extension.py", line 66, in run
      self.build_extension(ext)
    File "/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-build-env-nt4fesab/overlay/lib/python3.8/site-packages/cmake_build_extension/build_extension.py", line 173, in build_extension
      subprocess.check_call(configure_command)
    File "/Users/karsten/.pyenv/versions/3.8.0/Python.framework/Versions/3.8/lib/python3.8/subprocess.py", line 364, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['cmake', '-S', '/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-install-xinkxb_o/idyntree_6e7a01ce97a940e69151d9a2f521d9c0', '-B', '/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-install-xinkxb_o/idyntree_6e7a01ce97a940e69151d9a2f521d9c0/build/temp.macosx-10.14-x86_64-3.8_BuildAndInstall', '-GNinja', '-DCMAKE_BUILD_TYPE=Release', '-DCMAKE_INSTALL_PREFIX:PATH=/private/var/folders/0x/mb9v40950hl6630t8xm7y26c0000gn/T/pip-install-xinkxb_o/idyntree_6e7a01ce97a940e69151d9a2f521d9c0/build/lib.macosx-10.14-x86_64-3.8/idyntree', '-DBUILD_SHARED_LIBS:BOOL=OFF', '-DIDYNTREE_USES_PYTHON:BOOL=ON', '-DIDYNTREE_COMPILE_TESTS:BOOL=OFF', '-DIDYNTREE_PACKAGE_FOR_PYPI:BOOL=ON', '-DIDYNTREE_USES_IPOPT:BOOL=ON', '-DIDYNTREE_USES_ASSIMP:BOOL=ON', '-DIDYNTREE_USES_IRRLICHT:BOOL=OFF', '-DIDYNTREE_USES_QT5:BOOL=OFF', '-DIDYNTREE_USES_OSQPEIGEN:BOOL=OFF', '-DIDYNTREE_USES_ALGLIB:BOOL=OFF', '-DIDYNTREE_USES_WORHP:BOOL=OFF', '-DIDYNTREE_USES_YARP:BOOL=OFF', '-DIDYNTREE_USES_ICUB_MAIN:BOOL=OFF']' returned non-zero exit status 1.
  ----------------------------------------
  ERROR: Failed building wheel for idyntree
Failed to build idyntree
ERROR: Could not build wheels for idyntree which use PEP 517 and cannot be installed directly
```

**Comment by traversaro** [comment]
Hi @Karsten1987, 
  we should indeed refresh a bit the README to document properly the way in which the library can be installed, including pip and conda. 

> Is the pip package supported on OSX?

Short answer: no, on macOS and Windows we recommend to use conda binary packages, especially when one is interested in Python bindings.

Long answer:

In a nutshell, if you just want to quickly experiment with Python bindings on Linux/macOS/Windows and Python 3.6/3.7/3.8/3.9, the easiest option is to use conda and install idyntree via conda:
~~~
conda install -c conda-forge -c robotology idyntree
~~~
This provides pre-compiled binaries for all the combinations of Linux/macOS/Windows and Python 3.6/3.7/3.8/3.9, avoiding the user for CMake configuration errors and compilation errors such as the one that you are experiencing. 

On the pip side, instead on Linux binaries are provided, while on Windows and macOS the source distribution is used, so the possibility of having problems is high, as you need to have all the dependencies installed on your own with some kind of package manager or by manually installing them, so those configuration are not explicitly supported neither in documentation nor in CI. However, in your problem seems to be happening in `find_package(pybind11 QUIET)`, and that line actually should not be run at all. I tried to provide a fix in https://github.com/robotology/idyntree/pull/877, I guess you could try it with:
~~~
pip install git+https://github.com/robotology/idyntree@fix876
~~~

**Comment by traversaro** [comment]
See the previous answer regarding the specific of installing with pip on macOS . However, I wanted to ask something related to your first sentence:

> We're trying out the library in the context of ros2_control but having trouble to install the library through pip.

Cool, that is quite interesting, do you have any link/reference? Are you interested to use the C++ API or the Python bindings? If you want to use the C++ API, there is any reason why you want to install the library via pip as opposed to building from source of package idyntree in whatever system you are using for dependencies?

**Comment by traversaro** [comment]
> I tried to provide a fix in #877, I guess you could try it with:
> 
> ```
> pip install git+https://github.com/robotology/idyntree@fix876
> ```

The PR was merged in master, so it should be sufficient to try as:
> ```
> pip install git+https://github.com/robotology/idyntree@master
> ```

**Comment by traversaro** [comment]
@Karsten1987 I think the issue should be closed, feel free to re-open if it is the case.
